pkg/validator/promql_expression.go

```diff
@@ -121,6 +121,47 @@ func (h expressionDoesNotUseLabels) Validate(_ unmarshaler.RuleGroup, rule rulef
 	return errs
 }
 
+type expressionUsesOnlyAllowedLabelsForMetricRegexp struct {
+	allowedLabels    []string
+	metricNameRegexp string
+}
+
+func newExpressionUseOnlyWhitelistedLabelsForMetric(paramsConfig yaml.Node) (Validator, error) {
+	params := struct {
+		AllowedLabels    []string `yaml:"allowedLabels"`
+		MetricNameRegexp string   `yaml:"metricNameRegexp"`
+	}{}
+	if err := paramsConfig.Decode(&params); err != nil {
+		return nil, err
+	}
+	if len(params.AllowedLabels) == 0 {
+		return nil, fmt.Errorf("missing labels")
+	}
+	return &expressionUsesOnlyAllowedLabelsForMetricRegexp{allowedLabels: params.AllowedLabels, metricNameRegexp: params.MetricNameRegexp}, nil
+}
+
+func (h expressionUsesOnlyAllowedLabelsForMetricRegexp) String() string {
+	return fmt.Sprintf("expression only uses allowed labels `%s` for metric %s", strings.Join(h.allowedLabels, "`,`"), h.metricNameRegexp)
+}
+
+func (h expressionUsesOnlyAllowedLabelsForMetricRegexp) Validate(_ unmarshaler.RuleGroup, rule rulefmt.Rule, _ *prometheus.Client) []error {
+	usedLabels, err := getExpressionUsedLabelsForMetricRegexp(rule.Expr, h.metricNameRegexp)
+	if err != nil {
+		return []error{err}
+	}
+	allowedLabelsMap := map[string]struct{}{}
```

Comment:
What about making already the `expressionUsesOnlyAllowedLabelsForMetricRegexp.allowedLabels` a `map[string]struct{}{}` so this happens only once during config unmarshalling and not on every validation?

Reply:
Refactored as proposed.